File: crates/instrument/src/classes/dmm.rs
use super::dialect_io;
use instrument_core::error::Result;
use instrument_core::kind::InstrumentKind;
use instrument_core::scpi::ScpiSession;
use instrument_core::scpi_commands;
use instrument_core::InstrumentSession;

/// Digital multimeter session view (IVI-inspired / SCPI :MEASure).
pub struct Dmm {
    session: InstrumentSession,
}

impl Dmm {
    pub fn new(session: InstrumentSession) -> Self {
        Self { session }
    }

    pub fn from_session(_session: &InstrumentSession) -> Result<Self> {
        Err(instrument_core::Error::Unsupported(
            "use SessionPool for shared session views",
        ))
    }

    pub fn session(&self) -> &InstrumentSession {
        &self.session
    }

    pub fn session_mut(&mut self) -> &mut InstrumentSession {
        &mut self.session
    }

    fn dialect(&self) -> &'static instrument_core::DialectProfile {
        self.session.dialect_for(InstrumentKind::Dmm)
    }

    fn cmd(&self, key: &str, vars: &[(&str, String)], fallback: String) -> String {
        dialect_io::try_formatted(self.dialect(), key, vars, fallback)
    }

    /// Measures DC voltage in volts (SI).
    pub fn measure_voltage_dc(&mut self, range: Option<f64>) -> Result<f64> {
        let cmd = self.cmd(
            "measure_voltage_dc",
            &dialect_io::range_vars(range),
            scpi_commands::dmm_measure_voltage_dc(range),
        );
        self.query_f64(&cmd)
    }

    /// Measures AC voltage in volts (SI).
    pub fn measure_voltage_ac(&mut self, range: Option<f64>) -> Result<f64> {
        let cmd = self.cmd(
            "measure_voltage_ac",
            &dialect_io::range_vars(range),
            scpi_commands::dmm_measure_voltage_ac(range),
        );
        self.query_f64(&cmd)
    }

    /// Measures DC current in amps (SI).
    pub fn measure_current_dc(&mut self, range: Option<f64>) -> Result<f64> {
        let cmd = self.cmd(
            "measure_current_dc",
            &dialect_io::range_vars(range),
            scpi_commands::dmm_measure_current_dc(range),
        );
        self.query_f64(&cmd)
    }

    /// Measures AC current in amps (SI).
    pub fn measure_current_ac(&mut self, range: Option<f64>) -> Result<f64> {
        let cmd = self.cmd(
            "measure_current_ac",
            &dialect_io::range_vars(range),
            scpi_commands::dmm_measure_current_ac(range),
        );
        self.query_f64(&cmd)
    }

    /// Measures 2-wire resistance in ohms (SI).
    pub fn measure_resistance(&mut self, range: Option<f64>) -> Result<f64> {
        self.measure_resistance_2wire(range)
    }

    /// Measures 2-wire resistance in ohms (SI).
    pub fn measure_resistance_2wire(&mut self, range: Option<f64>) -> Result<f64> {
        let cmd = self.cmd(
            "measure_resistance_2w",
            &dialect_io::range_vars(range),
            scpi_commands::dmm_measure_resistance_2wire(range),
        );
        self.query_f64(&cmd)
    }

    /// Measures 4-wire resistance in ohms (SI).
    pub fn measure_resistance_4wire(&mut self, range: Option<f64>) -> Result<f64> {
        let cmd = self.cmd(
            "measure_resistance_4w",
            &dialect_io::range_vars(range),
            scpi_commands::dmm_measure_resistance_4wire(range),
        );
        self.query_f64(&cmd)
    }

    /// Measures temperature (instrument units / °C when configured as Celsius).
    pub fn measure_temperature(&mut self, range: Option<f64>) -> Result<f64> {
        let cmd = self.cmd(
            "measure_temperature",
            &dialect_io::range_vars(range),
            scpi_commands::dmm_measure_temperature(range),
        );
        self.query_f64(&cmd)
    }

    /// Configures DC voltage measurement for faster repeated reads.
    pub fn configure_voltage_dc(
        &mut self,
        range: Option<f64>,
        resolution: Option<f64>,
    ) -> Result<()> {
        let cmd = self.cmd(
            "configure_voltage_dc",
            &dialect_io::range_resolution_vars(range, resolution),
            scpi_commands::dmm_configure_voltage_dc(range, resolution),
        );
        self.session.scpi_mut().write(&cmd)
    }

    /// Configures AC voltage measurement for faster repeated reads.
    pub fn configure_voltage_ac(
        &mut self,
        range: Option<f64>,
        resolution: Option<f64>,
    ) -> Result<()> {
        let cmd = self.cmd(
            "configure_voltage_ac",
            &dialect_io::range_resolution_vars(range, resolution),
            scpi_commands::dmm_configure_voltage_ac(range, resolution),
        );
        self.session.scpi_mut().write(&cmd)
    }

    /// Configures DC current measurement for faster repeated reads.
    pub fn configure_current_dc(
        &mut self,
        range: Option<f64>,
        resolution: Option<f64>,
    ) -> Result<()> {
        let cmd = self.cmd(
            "configure_current_dc",
            &dialect_io::range_resolution_vars(range, resolution),
            scpi_commands::dmm_configure_current_dc(range, resolution),
        );
        self.session.scpi_mut().write(&cmd)
    }

    /// Configures AC current measurement for faster repeated reads.
    pub fn configure_current_ac(
        &mut self,
        range: Option<f64>,
        resolution: Option<f64>,
    ) -> Result<()> {
        let cmd = self.cmd(
            "configure_current_ac",
            &dialect_io::range_resolution_vars(range, resolution),
            scpi_commands::dmm_configure_current_ac(range, resolution),
        );
        self.session.scpi_mut().write(&cmd)
    }

    /// Configures 2-wire resistance measurement for faster repeated reads.
    pub fn configure_resistance(
        &mut self,
        range: Option<f64>,
        resolution: Option<f64>,
    ) -> Result<()> {
        let cmd = self.cmd(
            "configure_resistance",
            &dialect_io::range_resolution_vars(range, resolution),
            scpi_commands::dmm_configure_resistance(range, resolution),
        );
        self.session.scpi_mut().write(&cmd)
    }

    /// Configures 4-wire resistance measurement for faster repeated reads.
    pub fn configure_resistance_4wire(
        &mut self,
        range: Option<f64>,
        resolution: Option<f64>,
    ) -> Result<()> {
        let cmd = self.cmd(
            "configure_resistance_4w",
            &dialect_io::range_resolution_vars(range, resolution),
            scpi_commands::dmm_configure_resistance_4wire(range, resolution),
        );
        self.session.scpi_mut().write(&cmd)
    }

    /// Initiates a measurement (INIT).
    pub fn initiate(&mut self) -> Result<()> {
        let cmd = dialect_io::try_command(self.dialect(), "initiate", scpi_commands::DMM_INITIATE);
        self.session.scpi_mut().write(cmd)
    }

    /// Fetches the last initiated measurement (FETC?).
    pub fn fetch(&mut self) -> Result<f64> {
        self.query_f64(dialect_io::try_command(
            self.dialect(),
            "fetch",
            scpi_commands::DMM_FETCH,
        ))
    }

    /// Reads a measurement immediately (READ?).
    pub fn read(&mut self) -> Result<f64> {
        self.query_f64(dialect_io::try_command(
            self.dialect(),
            "read",
            scpi_commands::DMM_READ,
        ))
    }

    /// Issues a software trigger (*TRG).
    pub fn software_trigger(&mut self) -> Result<()> {
        let cmd = dialect_io::try_command(
            self.dialect(),
            "software_trigger",
            scpi_commands::DMM_SOFTWARE_TRIGGER,
        );
        self.session.scpi_mut().write(cmd)
    }

    fn query_f64(&mut self, cmd: &str) -> Result<f64> {
        let resp = self.session.scpi_mut().query(cmd)?;
        ScpiSession::parse_f64(&resp)
    }
}

File: dotnet/src/InstrumentComponents/Classes/Dmm.cs
using InstrumentComponents.Dialects;
using InstrumentComponents.Identity;
using InstrumentComponents.Kind;
using InstrumentComponents.Session;
using InstrumentComponents.Scpi;

namespace InstrumentComponents.Classes;

/// <summary>Digital multimeter session view (IVI-inspired / SCPI :MEASure).</summary>
public sealed class Dmm : IInstrumentIdentity, IInstrumentShutdown
{
    private readonly InstrumentSession _session;

    public Dmm(InstrumentSession session) => _session = session;

    public InstrumentSession Session => _session;

    public Idn QueryIdn() => _session.Idn();

    public void Reset() => _session.Reset();

    /// <summary>DMMs have no output stage; safe-shutdown is a no-op.</summary>
    public void OutputOff()
    {
    }

    private DialectProfile Dialect => _session.DialectFor(InstrumentKind.Dmm);

    private string Cmd(string key, string fallback, params (string Name, string Value)[] vars) =>
        DialectCommand.Try(Dialect, key, fallback, vars);

    public double MeasureVoltageDc(double? range = null) =>
        QueryF64(Cmd("measure_voltage_dc", ScpiCommands.DmmMeasureVoltageDc(range), DialectCommand.RangeVars(range)));

    public double MeasureVoltageAc(double? range = null) =>
        QueryF64(Cmd("measure_voltage_ac", ScpiCommands.DmmMeasureVoltageAc(range), DialectCommand.RangeVars(range)));

    public double MeasureCurrentDc(double? range = null) =>
        QueryF64(Cmd("measure_current_dc", ScpiCommands.DmmMeasureCurrentDc(range), DialectCommand.RangeVars(range)));

    public double MeasureCurrentAc(double? range = null) =>
        QueryF64(Cmd("measure_current_ac", ScpiCommands.DmmMeasureCurrentAc(range), DialectCommand.RangeVars(range)));

    public double MeasureResistance(double? range = null) =>
        MeasureResistance2Wire(range);

    public double MeasureResistance2Wire(double? range = null) =>
        QueryF64(Cmd("measure_resistance_2w", ScpiCommands.DmmMeasureResistance2wire(range), DialectCommand.RangeVars(range)));

    public double MeasureResistance4Wire(double? range = null) =>
        QueryF64(Cmd("measure_resistance_4w", ScpiCommands.DmmMeasureResistance4wire(range), DialectCommand.RangeVars(range)));

    public double MeasureTemperature(double? range = null) =>
        QueryF64(Cmd("measure_temperature", ScpiCommands.DmmMeasureTemperature(range), DialectCommand.RangeVars(range)));

    public void ConfigureVoltageDc(double? range = null, double? resolution = null) =>
        _session.Scpi.Write(Cmd("configure_voltage_dc", ScpiCommands.DmmConfigureVoltageDc(range, resolution), DialectCommand.RangeResolutionVars(range, resolution)));

    public void ConfigureVoltageAc(double? range = null, double? resolution = null) =>
        _session.Scpi.Write(Cmd("configure_voltage_ac", ScpiCommands.DmmConfigureVoltageAc(range, resolution), DialectCommand.RangeResolutionVars(range, resolution)));

    public void ConfigureCurrentDc(double? range = null, double? resolution = null) =>
        _session.Scpi.Write(Cmd("configure_current_dc", ScpiCommands.DmmConfigureCurrentDc(range, resolution), DialectCommand.RangeResolutionVars(range, resolution)));

    public void ConfigureCurrentAc(double? range = null, double? resolution = null) =>
        _session.Scpi.Write(Cmd("configure_current_ac", ScpiCommands.DmmConfigureCurrentAc(range, resolution), DialectCommand.RangeResolutionVars(range, resolution)));

    public void ConfigureResistance(double? range = null, double? resolution = null) =>
        _session.Scpi.Write(Cmd("configure_resistance", ScpiCommands.DmmConfigureResistance(range, resolution), DialectCommand.RangeResolutionVars(range, resolution)));

    public void ConfigureResistance4Wire(double? range = null, double? resolution = null) =>
        _session.Scpi.Write(Cmd("configure_resistance_4w", ScpiCommands.DmmConfigureResistance4wire(range, resolution), DialectCommand.RangeResolutionVars(range, resolution)));

    public void Initiate() => _session.Scpi.Write(Cmd("initiate", ScpiCommands.DmmInitiate));

    public double Fetch() => QueryF64(Cmd("fetch", ScpiCommands.DmmFetch));

    public double Read() => QueryF64(Cmd("read", ScpiCommands.DmmRead));

    public void SoftwareTrigger() => _session.Scpi.Write(Cmd("software_trigger", ScpiCommands.DmmSoftwareTrigger));

    private double QueryF64(string cmd) =>
        ScpiSession.ParseF64(_session.Scpi.Query(cmd));
}
